File: Algorithm/Solution/2021-01-14/송원석/MergeSort.java
package CSstudy;

import java.util.Arrays;

public class MergeSort {
	public static void merge(int[] array, int left, int mid, int right) {
		int[] leftArray = Arrays.copyOfRange(array, left, mid+1);
		int[] rightArray = Arrays.copyOfRange(array, mid, right+1);
		
		int leftIdx = 0;
		int rightIdx = 0;
		
		for(int i=left;i<=right;i++) {
			if(rightIdx >= rightArray.length || leftArray[leftIdx] < rightArray[rightIdx]) {
				array[i] = leftArray[leftIdx++];
			}else{
				array[i] = rightArray[rightIdx++];
			}
		}
	}
	
	public static void mergeSort(int[] array, int left, int right) {
		if(left >= right) return;
		
		int mid = (left+right)/2;
		
		mergeSort(array, left, mid);
		mergeSort(array, mid+1, right);
		
		merge(array, left, mid, right);
		
	}

	public static void main(String[] args) {
		int[] array = {5,3,9,4,1,6,2,7};
		mergeSort(array, 0, array.length-1);
		
		System.out.println(Arrays.toString(array));
	}

}

File: Algorithm/Solution/2021-01-14/송원석/MergeSort.js
const merge = (array, left, mid, right) => {
  const leftArray = array.slice(left, mid + 1);
  const rightArray = array.slice(mid + 1, right + 1);
  let leftIndex = 0;
  let rightIndex = 0;
  for (let i = left; i <= right; i++) {
    if (leftIndex >= leftArray.length) {
      array[i] = rightArray[rightIndex++];
    } else if (rightIndex >= rightArray.length) {
      array[i] = leftArray[leftIndex++];
    } else if (leftArray[leftIndex] < rightArray[rightIndex]) {
      array[i] = leftArray[leftIndex++];
    } else {
      array[i] = rightArray[rightIndex++];
    }
  }
}


const mergeSort = (array, left, right) => {
  if (left < right) {
    const mid = Math.floor((left + right) / 2);
    mergeSort(array, left, mid);
    mergeSort(array, mid + 1, right);
    merge(array, left, mid, right);
  }
}


// const array = [3, 2, 1, 6, 10, 21];
const array = [5, 3, 8, 4, 9, 1, 6, 2, 7]
mergeSort(array, 0, array.length - 1);

console.log(array);